### PR TITLE
Document upstream plugin registry changes

### LINKED ISSUE #18793: Impact of new plugin model on the documentation
We need changes to the documentation reflection changes in the new plugin model.

For example, this procedure is broken:  https://www.eclipse.org/che/docs/che-7/contributor-guide/branding-che-theia/

### PR DESCRIPTION
**Pull request by ericwill**
Signed-off-by: Eric Williams <ericwill@redhat.com>

> Read our [Contribution guide](https://github.com/eclipse/che-docs/blob/master/CONTRIBUTING.adoc) before submitting a PR.

### What does this PR do?
* Add details about adding che-theia-plugins.yaml entries on the page
  "Building custom registries"
* Contribute a new upstream specific article on how to write che-theia-plugins.yaml
  entries
* Small fixes to other articles
* Restore deps needed for local development to package.json

### What issues does this PR fix or reference?
Fixes eclipse/che#18793

### Specify the version of the product this PR applies to.
7.24.x and higher

### PR Checklist

As the author of this Pull Request I made sure that:

- [x] `vale` has been run successfully against the PR branch
- [x] Link checker has been run successfully against the PR branch
- [x] Documentation describes a scenario that is already covered by QE tests, otherwise an issue has been created and acknowledged by Che QE team
- [ ] Changed article references are updated where they are used (or a redirect has been set up on the docs side):
    - [ ] Dashboard [branding.json](https://github.com/eclipse/che-dashboard/blob/master/src/components/branding/branding.json)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)

### CODE REVIEW

File: modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc [hunk 1]
```diff
@@ -0,0 +1,12 @@
+. Observe the contents of `./{che-plugin-registry-directory}/v3/plugins/` present in the container after building the registry. All `meta.yaml` files resulting from a successful plugin registry build will be located here.
++
+[subs="+attributes,quotes"]
+----
+./{che-plugin-registry-directory}/v3/plugins/
```
Comment (tolusha) [inline modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc:5]: As far as I understand initially that directory is empty, right?
https://github.com/eclipse/che-plugin-registry/tree/master/v3/plugins

File: modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+.Procedure
+
+. Edit the `che-theia-plugins.yaml` file and create a new entry.
+
++
+[source,yaml]
+----
+- id: publisher/my-vscode-ext                                    <1>
+- repository:                                                    <2>
```
Comment (tolusha) [inline modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc]:
```suggestion
  repository:                                                    <2>
```

?

File: modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+<16> The name of the mount
+<17> The path of the mount
+<18> (OPTIONAL) Any endpoint information for the sidecar container
+<19> Endpoint name
+<20> Whether or not the endpoint is exposed publically or not
+<21> The port number
+<22> Attributes relating to the endpoint
+<23> Direct link(s) to the vsix files included with this plugin. The vsix built by the repository specified (i.e. the "main extension) must be listed first
+
+. Run the `build.sh` script with the options of your choosing. The build process will generate `meta.yaml` files automatically, based on the entries in the `che-theia-plugins.yaml` file.
```
Comment (tolusha) [inline modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc:83]: I am not sure if it is related to the PR but build failed for me
```
[test] Summary of all failing tests
[test]  FAIL  tests/sidecar/sidecar-docker-image.spec.ts
  ● Test Sidecar › basics
[test] 
    TypeError: Object.fromEntries is not a function

      47 |       n: '1',
      48 |     };
    > 49 |     const result = await this.git.log(logOptions);
         |                                   ^
      50 |     const latest = result.latest;
      51 |     if (!latest) {
      52 |       throw new Error(`Unable to find result when executing ${JSON.stringify(logOptions)}`);
[test] 
      at userOptions (../../node_modules/simple-git/src/lib/tasks/log.ts:63:18)
      at parseLogOptions (../../node_modules/simple-git/src/lib/tasks/log.ts:102:22)
[test]       at Git.Object.<anonymous>.Git.log (../../node_modules/simple-git/src/git.js:871:26)
      at SidecarDockerImage.<anonymous> (src/sidecar/sidecar-docker-image.ts:49:35)
[test]       at src/sidecar/sidecar-docker-image.ts:1824:41
[test]       at Object.<anonymous>.__awaiter (src/sidecar/sidecar-docker-image.ts:1767:10)
[test]       at SidecarDockerImage.getDockerImageFor (src/sidecar/sidecar-docker-image.ts:1878:12)
[test]       at tests/sidecar/sidecar-docker-image.spec.ts:32:45
[test] 
[test]  FAIL  tests/registry/registry-helper.spec.ts
[test]   ● Test RegistryHelper › basics with current sha1
[test] 
[test]     TypeError: Object.fromEntries is not a function
[test] 
[test]       39 |       n: '1',
[test]       40 |     };
[test]     > 41 |     const result = await this.git.log(logOptions);
[test]          |                                   ^
[test]       42 |     // eslint-disable-next-line @typescript-eslint/no-explicit-any
[test]       43 |     const hash = (result.latest as any).hash;
[test]       44 |     this.shortSha1 = hash.substring(0, 7);
[test] 
[test]       at userOptions (../../node_modules/simple-git/src/lib/tasks/log.ts:63:18)
[test]       at parseLogOptions (../../node_modules/simple-git/src/lib/tasks/log.ts:102:22)
[test]       at Git.Object.<anonymous>.Git.log (../../node_modules/simple-git/src/git.js:871:26)
[test]       at RegistryHelper.<anonymous> (src/registry/registry-helper.ts:41:35)
      at fulfilled (src/registry/registry-helper.ts:2196:24)
[test] 
[test]  FAIL  tests/che-theia-plugin/che-theia-plugin-analyzer.spec.ts
[test]   ● Test CheTheiaPluginsAnalyzer › basics
[test] 
[test]     TypeError: Cannot read property 'url' of undefined
[test] 
[test]       44 | 
[test]       45 |     // search without id as it is not provided
[test]     > 46 |     const goPlugins = result.plugins.filter(plugin => plugin.repository.url === 'https://github.com/golang/vscode-go');
[test]          |                                                                         ^
[test]       47 |     expect(goPlugins).toBeDefined();
[test]       48 |     expect(goPlugins.length).toBe(1);
[test]       49 |     const goPlugin = goPlugins[0];
[test] 
[test]       at result.plugins.filter.plugin (tests/che-theia-plugin/che-theia-plugin-analyzer.spec.ts:46:73)
[test]           at Array.filter (<anonymous>)
[test]       at tests/che-theia-plugin/che-theia-plugin-analyzer.spec.ts:46:38
[test]       at fulfilled (tests/che-theia-plugin/che-theia-plugin-analyzer.spec.ts:5:58)
[test] 
[test] 
[test] Test Suites: 3 failed, 25 passed, 28 total
[test] Tests:       3 failed, 93 passed, 96 total
[test] Snapshots:   0 total
[test] Time:        8.109 s, estimated 13 s
[test] Ran all test suites.
[test] Force exiting Jest: Have you considered using `--detectOpenHandles` to detect async operations that kept running after all tests finished?
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[test] yarn test exited with code 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Comment (benoitf) [reply modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc:83]: probably old nodejs version ?

**Comment by MichalMaler** [comment]
Hello. Let me know when you will be ready with your edits and I will take a look :).

File: modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc (hunk 1 above)
Comment (ericwill) [reply modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc:5]: Correct, but this specific snippet is about post-build output. At that point the folder should have content in it.

**Comment by themr0c** [comment]
For 7.24.x and higher?

**Comment by ericwill** [comment]
> For 7.24.x and higher?

Yes, this PR adds upstream only docs for things that were part of Che 7.24 already. But no backporting is needed as these are only upstream docs.

File: modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc (hunk 1 above)
Comment (themr0c) [reply modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc:5]: This is probably a leftover from the time we used the same example with variables for upstream and downstream.

**Comment by themr0c** [comment]
Backporting to 7.24.x won't be possible, because the other changes for customizing the registries were not backported there.

Backporting to 7.26.x may be necessary if it introduces changes for CRW 2.7.

When there as as many ifeval statement in a file, it's probably a good indicator that the whole file should land in the examples directory as a context specific procedure.

File: modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc (hunk 1 above)
Comment (ericwill) [reply modules/administration-guide/examples/snip_che-plugin-registry-build-output.adoc:5]: This is a new file that I added, as it's specific to this PR's contents.

**Comment by ericwill** [comment]
> Backporting to 7.24.x won't be possible, because the other changes for customizing the registries were not backported there.
> 
> Backporting to 7.26.x may be necessary if it introduces changes for CRW 2.7.
> 
> When there as as many ifeval statement in a file, it's probably a good indicator that the whole file should land in the examples directory as a context specific procedure.

We do not need to backport this change. It's upstream only and AFAIK we are not publishing multiple versions of the upstream docs. Since we do not care about this PR for downstream we do not need to backport.

File: modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc (hunk 2 above)
Comment (tolusha) [reply modules/end-user-guide/partials/proc_adding-a-vs-code-extension-to-the-che-plugin-registry.adoc]: @ericwill 
Is it typo ?

**Comment by MichalMaler** [comment]
Waiting for Eric @ericwill  (PTO) to answer the last question from Anatolii @tolusha

**Review by MichalMaler — APPROVED** [review]

**Comment by MichalMaler** [comment]
@benoitf @tolusha LGTM from a doc perspective. Can you resolve your conversations so we could get the last green mark and merge this? :) Thank you!

**Review by tolusha — APPROVED** [review]

**Comment by MichalMaler** [comment]
This PR is ready to be merged, but we need to put it on hold until the che `7.28.x` branch will be created. By doing so, we will be able to avoid downstreaming of this content to CRW where it doesn't belong.

**Pull request merged**